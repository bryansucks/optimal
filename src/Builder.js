/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import invariant from './invariant';
import isObject from './isObject';

import type { SupportedType, Checker } from './types';

export default class Builder<T> {
  checks: {
    args: *[],
    func: Checker,
  }[] = [];
  defaultValue: T;
  nullable: boolean = true;
  type: SupportedType;

  constructor(type: SupportedType, defaultValue: T) {
    if (typeof defaultValue === 'undefined') {
      throw new TypeError(`A default value for type "${type}" is required.`);
    }

    this.defaultValue = defaultValue;
    this.type = type;

    this.addCheck(this.checkTypeOf);
  }

  /**
   * Add a checking function with optional arguments.
   */
  addCheck(func: Checker, ...args: *[]): this {
    this.checks.push({
      args,
      func,
    });

    return this;
  }

  /**
   * Validate the value matches only the default value.
   */
  checkOnly(path: string, value: *) {
    invariant(
      (value === this.defaultValue),
      `Value may only be "${String(this.defaultValue)}".`,
      path,
    );
  }

  /**
   * Validate the type of value.
   */
  checkTypeOf(path: string, value: *) {
    switch (this.type) {
      case 'array':
        invariant(Array.isArray(value), 'Must be an array.', path);
        break;

      case 'object':
        invariant(isObject(value), 'Must be a plain object.', path);
        break;

      default:
        // eslint-disable-next-line valid-typeof
        invariant((typeof value === this.type), `Must be a ${this.type}.`, path);
        break;
    }
  }

  /**
   * Mark a field as only the default value can be used.
   */
  only(): this {
    invariant(
      // eslint-disable-next-line valid-typeof
      (typeof this.defaultValue === this.type),
      `only() requires a default ${this.type} value.`,
    );

    return this.addCheck(this.checkOnly);
  }

  /**
   * Mark a field as required and disallow nulls.
   */
  required(): this {
    this.nullable = false;

    return this;
  }

  /**
   * Run all validation checks that have been enqueued.
   */
  runChecks(path: string, initialValue: *): * {
    const value = (typeof initialValue === 'undefined') ? this.defaultValue : initialValue;

    // If nullable, just abort early
    if (value === null && this.nullable) {
      return value;
    }

    // Run all checks against the value
    this.checks.forEach((checker) => {
      checker.func.call(this, path, value, ...checker.args);
    });

    return value;
  }
}
