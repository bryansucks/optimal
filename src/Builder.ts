/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import isObject from './isObject';
import { SupportedType, CheckerCallback, CustomCallback, OptimalOptions, Struct } from './types';

export interface Check {
  args: any[];
  callback: CheckerCallback;
}

export default class Builder<T> {
  checks: Check[] = [];

  currentStruct: Struct = {};

  defaultValue: T;

  deprecatedMessage: string = '';

  errorMessage: string = '';

  isNullable: boolean = false;

  isRequired: boolean = false;

  options: OptimalOptions = {};

  type: SupportedType;

  constructor(type: SupportedType, defaultValue: T) {
    if (__DEV__) {
      this.invariant(
        typeof defaultValue !== 'undefined',
        `A default value for type "${type}" is required.`,
      );

      this.addCheck(this.checkType);
    }

    this.defaultValue = defaultValue;
    this.type = type;
  }

  /**
   * Add a checking function with optional arguments.
   */
  addCheck(checker: CheckerCallback, ...args: any[]): this {
    if (__DEV__) {
      this.checks.push({
        args,
        callback: checker,
      });
    }

    return this;
  }

  /**
   * Map a list of names that must be defined alongside this field.
   */
  and(...keys: string[]): this {
    if (__DEV__) {
      this.invariant(keys.length > 0, 'AND requires a list of field names.');
    }

    return this.addCheck(this.checkAnd, keys);
  }

  /**
   * Validate that all fields have been defined.
   */
  checkAnd(path: string, value: any, otherKeys: string[]) {
    if (__DEV__) {
      const keys = [this.key(path), ...otherKeys];
      const struct = this.currentStruct;
      const undefs = keys.filter(key => typeof struct[key] === 'undefined' || struct[key] === null);

      // Only error once one of the struct is defined
      if (undefs.length === keys.length) {
        return;
      }

      this.invariant(
        undefs.length === 0,
        `All of these fields must be defined: ${keys.join(', ')}`,
      );
    }
  }

  /**
   * Validate the type of value.
   */
  checkType(path: string, value: any) {
    if (__DEV__) {
      switch (this.type) {
        case 'array':
          this.invariant(Array.isArray(value), 'Must be an array.', path);
          break;

        case 'custom':
        case 'instance':
        case 'union':
          // Handle in the sub-class
          break;

        case 'object':
        case 'shape':
          this.invariant(isObject(value), 'Must be a plain object.', path);
          break;

        default:
          // eslint-disable-next-line valid-typeof
          this.invariant(typeof value === this.type, `Must be a ${this.type}.`, path);
          break;
      }
    }
  }

  /**
   * Set a callback to run custom logic.
   */
  custom(callback: CustomCallback): this {
    if (__DEV__) {
      this.invariant(
        typeof callback === 'function',
        'Custom blueprints require a validation function.',
      );
    }

    return this.addCheck(this.checkCustom, callback);
  }

  /**
   * Validate the value using a custom callback.
   */
  checkCustom(path: string, value: any, callback: CustomCallback) {
    if (__DEV__) {
      try {
        callback(value, this.currentStruct);
      } catch (error) {
        this.invariant(false, error.message, path);
      }
    }
  }

  /**
   * Set a message to log when this field is present.
   */
  deprecate(message: string): this {
    if (__DEV__) {
      this.invariant(
        typeof message === 'string' && !!message,
        'A non-empty string is required for deprecated messages.',
      );

      this.deprecatedMessage = message;
    }

    return this;
  }

  /**
   * Throw an error if the condition is falsy.
   */
  invariant(condition: boolean, message: string, path: string = '') {
    if (__DEV__) {
      if (condition) {
        return;
      }

      const { file, name } = this.options;
      const error = this.errorMessage || message;
      let prefix = '';

      if (path) {
        if (name) {
          prefix += `Invalid ${name} field "${path}"`;
        } else {
          prefix += `Invalid field "${path}"`;
        }
      } else if (name) {
        prefix += name;
      }

      if (file) {
        prefix += ` in ${file}`;
      }

      if (prefix) {
        throw new Error(`${prefix}. ${error}`);
      } else {
        throw new Error(error);
      }
    }
  }

  /**
   * Return the current key from a path.
   */
  key(path: string): string {
    const index = path.lastIndexOf('.');

    return index > 0 ? path.slice(index + 1) : path;
  }

  /**
   * Set a custom error message for all checks.
   */
  message(message: string): this {
    if (__DEV__) {
      this.invariant(
        typeof message === 'string' && !!message,
        'A non-empty string is required for custom messages.',
      );

      this.errorMessage = message;
    }

    return this;
  }

  /**
   * Allow null values.
   */
  nullable(state: boolean = true): this {
    if (__DEV__) {
      this.isNullable = state;
    }

    return this;
  }

  /**
   * Mark a field as only the default value can be used.
   */
  only(): this {
    if (__DEV__) {
      this.invariant(
        // eslint-disable-next-line valid-typeof
        typeof this.defaultValue === this.type,
        `Only requires a default ${this.type} value.`,
      );
    }

    return this.addCheck(this.checkOnly);
  }

  /**
   * Validate the value matches only the default value.
   */
  checkOnly(path: string, value: any) {
    if (__DEV__) {
      this.invariant(
        value === this.defaultValue,
        `Value may only be "${String(this.defaultValue)}".`,
        path,
      );
    }
  }

  /**
   * Map a list of field names that must have at least 1 defined.
   */
  or(...keys: string[]): this {
    if (__DEV__) {
      this.invariant(keys.length > 0, 'OR requires a list of field names.');
    }

    return this.addCheck(this.checkOr, keys);
  }

  /**
   * Validate that at least 1 field is defined.
   */
  checkOr(path: string, value: any, otherKeys: string[]) {
    if (__DEV__) {
      const keys = [this.key(path), ...otherKeys];
      const struct = this.currentStruct;
      const defs = keys.filter(key => typeof struct[key] !== 'undefined' && struct[key] !== null);

      this.invariant(
        defs.length > 0,
        `At least one of these fields must be defined: ${keys.join(', ')}`,
      );
    }
  }

  /**
   * Disallow undefined values.
   */
  required(state: boolean = true): this {
    if (__DEV__) {
      this.isRequired = state;
    }

    return this;
  }

  /**
   * Run all validation checks that have been enqueued.
   */
  runChecks(path: string, initialValue: any, struct: Struct, options: OptimalOptions = {}): any {
    this.currentStruct = struct;
    this.options = options;

    let value = initialValue;

    // Handle undefined
    if (typeof value === 'undefined') {
      if (!this.isRequired) {
        value = this.defaultValue;
      } else if (__DEV__) {
        this.invariant(false, 'Field is required and must be defined.', path);
      }
    } else if (this.deprecatedMessage) {
      if (__DEV__) {
        // eslint-disable-next-line no-console
        console.info(`Field "${path}" is deprecated. ${this.deprecatedMessage}`);
      }
    }

    // Handle null
    if (value === null) {
      if (this.isNullable) {
        return value;
      }

      if (__DEV__) {
        this.invariant(false, 'Null is not allowed.', path);
      }
    }

    // Run all checks against the value
    if (__DEV__) {
      this.checks.forEach(checker => {
        checker.callback.call(this, path, value, ...checker.args);
      });
    }

    return value;
  }

  /**
   * Return a human readable type name.
   */
  typeAlias(): string {
    return this.type;
  }

  /**
   * Map a list of field names that must not be defined alongside this field.
   */
  xor(...keys: string[]): this {
    if (__DEV__) {
      this.invariant(keys.length > 0, 'XOR requires a list of field names.');
    }

    return this.addCheck(this.checkXor, keys);
  }

  /**
   * Validate that only 1 field is defined.
   */
  checkXor(path: string, value: any, otherKeys: string[]) {
    if (__DEV__) {
      const keys = [this.key(path), ...otherKeys];
      const struct = this.currentStruct;
      const defs = keys.filter(key => typeof struct[key] !== 'undefined' && struct[key] !== null);

      this.invariant(
        defs.length === 1,
        `Only one of these fields may be defined: ${keys.join(', ')}`,
      );
    }
  }
}

export function bool(defaultValue: boolean | null = false): Builder<boolean | null> {
  return new Builder('boolean', defaultValue);
}

export function custom<T>(
  callback: CustomCallback,
  defaultValue: T | null = null,
): Builder<T | null> {
  return new Builder('custom', defaultValue).custom(callback);
}

export function func(defaultValue: Function | null = null): Builder<Function | null> {
  return new Builder('function', defaultValue).nullable();
}
