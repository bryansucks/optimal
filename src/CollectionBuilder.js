/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';

export default class CollectionBuilder<T, TDefault> extends Builder<?TDefault> {
  contents: ?Builder<T>;

  constructor(
    type: 'array' | 'object',
    contents?: Builder<T>,
    defaultValue?: ?TDefault = null,
  ) {
    super(type, defaultValue);

    if (contents) {
      if (contents instanceof Builder) {
        this.contents = contents;
        this.addCheck(this.checkContents, contents);

      } else if (__DEV__) {
        this.invariant(false, `A blueprint is required for ${type} contents.`);
      }
    }
  }

  checkContents(path: string, value: *, contents: Builder<T>) {
    if (__DEV__) {
      if (this.type === 'array') {
        value.forEach((item, i) => {
          contents.runChecks(
            `${path}[${i}]`,
            item,
            this.currentOptions,
            this.currentConfig,
          );
        });

      } else if (this.type === 'object') {
        Object.keys(value).forEach((key) => {
          contents.runChecks(
            `${path}.${key}`,
            value[key],
            this.currentOptions,
            this.currentConfig,
          );
        });
      }
    }
  }

  notEmpty(): this {
    return this.addCheck(this.checkNotEmpty);
  }

  checkNotEmpty(path: string, value: *) {
    if (__DEV__) {
      if (this.type === 'array') {
        this.invariant(
          (value.length > 0),
          'Array cannot be empty.',
          path,
        );

      } else if (this.type === 'object') {
        this.invariant(
          (Object.keys(value).length > 0),
          'Object cannot be empty.',
          path,
        );
      }
    }
  }

  /**
   * If contents are defined, return the type name using generics syntax.
   */
  typeAlias(): string {
    const { contents } = this;
    const alias = super.typeAlias();

    return contents ? `${alias}<${contents.typeAlias()}>` : alias;
  }
}

export function array<T>(
  contents?: Builder<T>,
  defaultValue?: ?T[] = [],
): CollectionBuilder<T, T[]> {
  return new CollectionBuilder('array', contents, defaultValue);
}

export function object<T>(
  contents?: Builder<T>,
  defaultValue?: ?{ [key: string]: T } = {},
): CollectionBuilder<T, { [key: string]: T }> {
  return new CollectionBuilder('object', contents, defaultValue);
}
