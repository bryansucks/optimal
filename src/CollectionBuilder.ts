import Builder from './Builder';

export default class CollectionBuilder<T> extends Builder<T> {
  sizeOf(length: number): this {
    if (__DEV__) {
      this.invariant(
        typeof length === 'number' && length > 0,
        'Size requires a non-zero positive number.',
      );
    }

    return this.addCheck(this.checkSizeOf, length);
  }

  checkSizeOf(path: string, value: T, length: number) {
    if (__DEV__) {
      // Array
      if (Array.isArray(value)) {
        this.invariant(value.length === length, `Array length must be ${length}.`, path);

        // String
      } else if (typeof value === 'string') {
        this.invariant(value.length === length, `String length must be ${length}.`, path);

        // Object
      } else if (typeof value === 'object' && value) {
        this.invariant(
          Object.keys(value).length === length,
          `Object must have ${length} properties.`,
          path,
        );

        // Unknown
      } else {
        this.invariant(false, 'Unknown type for size of checks.');
      }
    }
  }
}
