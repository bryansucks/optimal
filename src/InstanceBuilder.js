/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';
import isObject from './isObject';

// eslint-disable-next-line flowtype/no-weak-types
export default class InstanceBuilder<T: Function> extends Builder<?T> {
  refClass: ?T;

  constructor(refClass?: ?T = null) {
    super('instance', null);

    // Nullable by default
    this.nullable();

    if (__DEV__) {
      if (refClass) {
        this.invariant(typeof refClass === 'function', 'A class reference is required.');
      }

      this.refClass = refClass;
      this.addCheck(this.checkInstance, refClass);
    }
  }

  checkInstance(path: string, value: *, refClass: ?T) {
    if (__DEV__) {
      if (refClass) {
        this.invariant(
          value instanceof refClass,
          `Must be an instance of "${this.typeAlias()}".`,
          path,
        );
      } else {
        this.invariant(
          isObject(value) && value.constructor !== Object,
          'Must be a class instance.',
          path,
        );
      }
    }
  }

  /**
   * If reference class is defined, return the class name if available.
   */
  typeAlias(): string {
    const { refClass } = this;

    return refClass ? refClass.name || refClass.constructor.name : 'Class';
  }
}

// eslint-disable-next-line flowtype/no-weak-types
export function instance<T: Function>(refClass?: ?T = null): InstanceBuilder<T> {
  return new InstanceBuilder(refClass);
}

export function regex(): InstanceBuilder<Class<RegExp>> {
  return instance(RegExp);
}

export function date(): InstanceBuilder<Class<Date>> {
  return instance(Date);
}
