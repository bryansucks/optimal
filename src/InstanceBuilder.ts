import Builder from './Builder';
import isObject from './isObject';
import instanceOf from './instanceOf';

export type Constructor<T> = new (...args: any[]) => T;

export default class InstanceBuilder<T> extends Builder<T | null> {
  loose: boolean = false;

  refClass: Constructor<T> | null = null;

  constructor(refClass: Constructor<T> | null = null, loose: boolean = false) {
    super('instance', null);

    // Nullable by default
    this.nullable();

    if (__DEV__) {
      if (refClass) {
        this.invariant(typeof refClass === 'function', 'A class reference is required.');
      }

      this.loose = loose;
      this.refClass = refClass;
      this.addCheck(this.checkInstance, refClass);
    }
  }

  checkInstance(path: string, value: T, refClass: T | null) {
    if (__DEV__) {
      if (refClass) {
        this.invariant(
          typeof refClass === 'function' &&
            (value instanceof refClass ||
              (this.loose && isObject(value) && instanceOf(value, refClass))),
          `Must be an instance of "${this.typeAlias()}".`,
          path,
        );
      } else {
        this.invariant(
          isObject(value) && value.constructor !== Object,
          'Must be a class instance.',
          path,
        );
      }
    }
  }

  /**
   * If reference class is defined, return the class name if available.
   */
  typeAlias(): string {
    const { refClass } = this;

    return refClass ? refClass.name || refClass.constructor.name : 'class';
  }
}

export function instance<T = Function>(
  refClass: Constructor<T> | null = null,
  loose?: boolean,
) /* infer */ {
  return new InstanceBuilder<T>(refClass, loose);
}

export function regex() /* infer */ {
  return instance(RegExp);
}

export function date() /* infer */ {
  return instance(Date);
}
