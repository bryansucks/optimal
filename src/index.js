/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { bool, custom, func } from './Builder';
import { array, object } from './CollectionBuilder';
import { instance, date, regex } from './InstanceBuilder';
import { number } from './NumberBuilder';
import { shape } from './ShapeBuilder';
import { string } from './StringBuilder';
import { union } from './UnionBuilder';

export {
  array,
  bool,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  union,
};
export { default } from './Options';
