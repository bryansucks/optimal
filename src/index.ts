/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Builder, { bool, custom, func } from './Builder';
import ArrayBuilder, { array } from './ArrayBuilder';
import InstanceBuilder, { instance, date, regex } from './InstanceBuilder';
import NumberBuilder, { number } from './NumberBuilder';
import ObjectBuilder, { object } from './ObjectBuilder';
import ShapeBuilder, { shape } from './ShapeBuilder';
import StringBuilder, { string } from './StringBuilder';
import UnionBuilder, { union } from './UnionBuilder';

export { array, bool, custom, date, func, instance, number, object, regex, shape, string, union };

export {
  Builder,
  ArrayBuilder,
  InstanceBuilder,
  NumberBuilder,
  ObjectBuilder,
  ShapeBuilder,
  StringBuilder,
  UnionBuilder,
};

export * from './types';

export default optimal;
