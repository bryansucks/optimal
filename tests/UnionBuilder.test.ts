import UnionBuilder, { union } from '../src/UnionBuilder';
import { bool, custom } from '../src/Builder';
import { array, object } from '../src/CollectionBuilder';
import { instance } from '../src/InstanceBuilder';
import { number } from '../src/NumberBuilder';
import { shape } from '../src/ShapeBuilder';
import { string } from '../src/StringBuilder';

describe('UnionBuilder', () => {
  let builder: UnionBuilder;

  class Foo {}
  class Bar {}

  beforeEach(() => {
    builder = union([
      array(string()),
      bool(true).only(),
      instance(Foo),
      number().between(0, 5),
      object(number()),
      string('foo').oneOf(['foo', 'bar', 'baz']),
    ]);
  });

  describe('constructor()', () => {
    it('errors if a non-array is not passed', () => {
      expect(() => {
        // @ts-ignore
        union('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if an empty array is passed', () => {
      expect(() => {
        union([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if an array with non-builders is passed', () => {
      expect(() => {
        // @ts-ignore
        union([123]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if a builder array is passed', () => {
      expect(() => {
        union([string()]);
      }).not.toThrowError('A non-empty array of blueprints are required for a union.');
    });

    it('sets default value', () => {
      expect(union([string()], 'bar').defaultValue).toBe('bar');
    });
  });

  describe('runChecks()', () => {
    it('errors if a unsupported type is used', () => {
      expect(() => {
        union([string(), number(), bool()]).runChecks('key', [], {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a nested union is used', () => {
      expect(() => {
        union([string('foo').oneOf(['foo', 'bar', 'baz']), union([number(), bool()])]).runChecks(
          'key',
          [],
          {},
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if an object and shape are used', () => {
      expect(() => {
        union([
          object(string()),
          shape({
            foo: string(),
            bar: number(),
          }),
        ]).runChecks('key', [], {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if the same builder type is used multiple times', () => {
      expect(() => {
        union([object(string()), object(number())]).runChecks('key', [], {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors with the class name for instance checks', () => {
      expect(() => {
        union([number(), instance(FormData)]).runChecks('key', {}, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs array check', () => {
      expect(() => {
        builder.runChecks('key', [123], {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs boolean check', () => {
      expect(() => {
        builder.runChecks('key', false, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs custom check', () => {
      expect(() => {
        union([
          string(),
          custom(value => {
            if (typeof value === 'number') {
              throw new TypeError('Encountered a number!');
            }
          }),
        ]).runChecks('key', 123, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs instance check', () => {
      expect(() => {
        builder.runChecks('key', new Bar(), {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs number check', () => {
      expect(() => {
        builder.runChecks('key', 10, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs object check', () => {
      expect(() => {
        builder.runChecks('key', { foo: 'foo' }, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs shape check', () => {
      expect(() => {
        union([
          shape({
            foo: string(),
            bar: number(),
          }),
        ]).runChecks(
          'key',
          {
            foo: 123,
          },
          {},
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs string check', () => {
      expect(() => {
        builder.runChecks('key', 'qux', {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs correctly for valid values', () => {
      expect(builder.runChecks('key', 'foo', {})).toBe('foo');
      expect(builder.runChecks('key', 3, {})).toBe(3);
      expect(builder.runChecks('key', true, {})).toBe(true);
    });
  });

  describe('typeAlias()', () => {
    it('returns all the available type aliases separated by pipes', () => {
      expect(union([string(), number(), bool()]).typeAlias()).toBe('string | number | boolean');
    });

    it('supports complex structures', () => {
      expect(builder.typeAlias()).toBe(
        'array<string> | boolean | Foo | number | object<number> | string',
      );
    });
  });
});
