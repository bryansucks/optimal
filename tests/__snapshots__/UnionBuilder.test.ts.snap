// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnionBuilder constructor() errors if a non-array is not passed 1`] = `"A non-empty array of blueprints are required for a union."`;

exports[`UnionBuilder constructor() errors if an array with non-builders is passed 1`] = `"A non-empty array of blueprints are required for a union."`;

exports[`UnionBuilder constructor() errors if an empty array is passed 1`] = `"A non-empty array of blueprints are required for a union."`;

exports[`UnionBuilder runChecks() errors if a nested union is used 1`] = `"Invalid field \\"key\\". Nested unions are not supported."`;

exports[`UnionBuilder runChecks() errors if a unsupported type is used 1`] = `"Invalid field \\"key\\". Type must be one of: string, number, boolean"`;

exports[`UnionBuilder runChecks() errors if an object and shape are used 1`] = `"Invalid field \\"key\\". Objects and shapes within the same union are not supported."`;

exports[`UnionBuilder runChecks() errors if the same builder type is used multiple times 1`] = `"Invalid field \\"key\\". Multiple instances of \\"object\\" are not supported."`;

exports[`UnionBuilder runChecks() errors with the class name for instance checks 1`] = `"Invalid field \\"key\\". Type must be one of: number, Buffer"`;

exports[`UnionBuilder runChecks() runs array check 1`] = `"Invalid field \\"key[0]\\". Must be a string."`;

exports[`UnionBuilder runChecks() runs boolean check 1`] = `"Invalid field \\"key\\". Value may only be \\"true\\"."`;

exports[`UnionBuilder runChecks() runs custom check 1`] = `"Invalid field \\"key\\". Encountered a number!"`;

exports[`UnionBuilder runChecks() runs instance check 1`] = `"Invalid field \\"key\\". Must be an instance of \\"Foo\\"."`;

exports[`UnionBuilder runChecks() runs number check 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`UnionBuilder runChecks() runs object check 1`] = `"Invalid field \\"key.foo\\". Must be a number."`;

exports[`UnionBuilder runChecks() runs shape check 1`] = `"Invalid field \\"key.foo\\". Must be a string."`;

exports[`UnionBuilder runChecks() runs string check 1`] = `"Invalid field \\"key\\". String must be one of: foo, bar, baz"`;
